### PR TITLE
Bare-metal reimaging with FOG

### PR DESCRIPTION
**Pull request by zmc**

### CODE REVIEW

File: teuthology/lock/cli.py [hunk 1]
```diff
@@ -144,21 +145,40 @@ def main(ctx):
...
                 machines_to_update.append(machine)
                 teuthology.provision.create_if_vm(
                     ctx,
                     misc.canonicalize_hostname(machine),
                 )
+        with teuthology.parallel.parallel() as p:
+            for machine in reimage_machines:
+                p.spawn(teuthology.provision.reimage, ctx, machine)
+        for machine in updatekeys_machines:
+            keys.do_update_keys([machine])
```
Comment (vasukulkarni) [inline teuthology/lock/cli.py]: does it need inventory update as well here for gui updates in pulpito?

File: teuthology/provision/fog.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        )
+        host_tasks = self.get_deploy_tasks()
+        for task in host_tasks:
+            timestamp = task['createdTime']
+            time_delta = (
+                datetime.utcnow() - datetime.strptime(
+                    timestamp, self.timestamp_format)
+            ).total_seconds()
+            # FIXME yay magic
+            if time_delta < 5:
```
Comment (vasukulkarni) [inline teuthology/provision/fog.py]: some more additional comments here please, what is <5 :)

File: teuthology/provision/fog.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+                    self.remote.connect()
+                    break
+                except (
+                    socket.error,
+                    SSHException,
+                    MaxWhileTries,
+                    EOFError,
+                ):
+                    pass
+        # cmd = "while [ ! -e '%s' ]; do sleep 5; done" % self._sentinel_path
```
Comment (vasukulkarni) [inline teuthology/provision/fog.py]: these comments can be deleted

File: docs/siteconfig.rst [hunk 4]
```diff
@@ -229,3 +229,10 @@ Here is a sample configuration with many of the options set and documented::
     # Settings for [nsupdate-web](https://github.com/zmc/nsupdate-web)
     # Used by the [libcloud](https://libcloud.apache.org/) backend
     nsupdate_url: http://nsupdate.front.sepia.ceph.com/update
+
+    # Settings for https://fogproject.org/
+    fog:
+      endpoint: http://fog.example.com/fog
+      api_token: your_api_token
+      user_token: your_user_token
+      machine_types: ['mira', 'smithi']
```
Comment (vasukulkarni) [inline docs/siteconfig.rst]: Not holding this PR, but some more high level info would be helpful.

File: teuthology/lock/cli.py (hunk 1 above)
Comment (zmc) [reply teuthology/lock/cli.py]: That's handled by the `internal.push_inventory` task

File: teuthology/provision/fog.py (hunk 2 above)
Comment (zmc) [reply teuthology/provision/fog.py]: Yeah - a more appropriate comment would be good

File: teuthology/nuke/__init__.py [hunk 5]
```diff
@@ -301,6 +302,13 @@ def nuke_helper(ctx, should_unlock):
         # does not check to ensure if the node is 'up'
         # we want to be able to nuke a downed node
         check_lock.check_lock(ctx, None, check_up=False)
+    status = get_status(host)
+    if status['machine_type'] in provision.fog.get_types():
+        status['os_type']
```
Comment (dmick) [inline teuthology/nuke/__init__.py]: dead code?

Comment (zmc) [reply teuthology/nuke/__init__.py]: Yep!

**Comment by zmc** [comment]
@vasu I've updated

**Comment by zmc** [comment]
test this please

**Comment by zmc** [comment]
@vasukulkarni I've updated

**Comment by vasukulkarni** [comment]
@zmc LGTM, I see it has DNM..

File: teuthology/provision/fog.py (hunk 2 above)
Comment (dmick) [reply teuthology/provision/fog.py]: for me it would be totally resolved with s/time_delta/time_delta_in_sec/

**Comment by dmick** [comment]
It all looks fine to me.

**Comment by zmc** [comment]
We've been running this branch for our nightlies for over a week. The `fog.py` module has 90% test coverage. The FOG server is having no problems at all keeping up with the load. I think it's time.

**Comment by djgalloway** [comment]
So excite.  Thanks @zmc!

**Pull request merged**